ReSharper.FSharp/src/FSharp.Psi.Features/src/Parsing/FSharpTreeBuilderBase.fs

```diff
@@ -117,32 +117,24 @@ type FSharpTreeBuilderBase(lexer, document: IDocument, lifetime, path: VirtualFi
     member x.AdvanceToTokenOrRangeEnd(tokenType: TokenNodeType, range: range) =
         x.AdvanceToTokenOrPos(tokenType, range.End)
 
-    member x.AdvanceToXmlDoc(xmlDoc: XmlDoc, maxPos: pos) =
-        x.AdvanceToStart(xmlDoc.Range)
-
-        // todo: add separate token type for xmlDoc comment?
-        let maxOffset = x.GetOffset(maxPos)
-        while x.CurrentOffset < maxOffset && not (x.TokenType == FSharpTokenType.LINE_COMMENT && startsWith "///" (x.Builder.GetTokenText())) do
-            x.AdvanceLexer()
+    member x.AdvanceToXmlDocOwner(xmlDoc: XmlDoc, expectedType: TokenNodeType, declarationRange: range) =
+        let range = xmlDocOwnerStartRange xmlDoc declarationRange
+        x.AdvanceToTokenOrRangeStart(expectedType, range)
 
     member x.AdvanceToTokenOrPos(tokenType: TokenNodeType, pos: pos) =
         let maxOffset = x.GetOffset(pos)
         while x.CurrentOffset < maxOffset && x.TokenType != tokenType && not x.Eof do
             x.AdvanceLexer()
 
-    // todo: wrong ranges (e.g. missing union case bar)
-    // todo wrong order:
-    // todo multiple blocks:
     member x.MarkXmlDocOwner(xmlDoc: XmlDoc, expectedType: TokenNodeType, declarationRange: range) =
-        let xmlDocRange = xmlDoc.Range
-        let declarationStart = declarationRange.Start
-
-        if xmlDoc.IsEmpty || Position.posLt declarationStart xmlDocRange.Start then
-            x.AdvanceToTokenOrRangeStart(expectedType, declarationRange)
-        else
-            x.AdvanceToXmlDoc(xmlDoc, declarationStart)
+        x.AdvanceToXmlDocOwner(xmlDoc, expectedType, declarationRange)
+        let mark = x.Mark()
+        if not xmlDoc.IsEmpty then x.MarkAndDone(xmlDoc.Range, ElementType.XML_DOC_BLOCK)
+        mark
 
-        x.Mark()
+    member x.MarkOrMarkXmlDocOwner(xmlDoc: XmlDoc, declarationRange: range) =
```

Comment:
Is this a typo in the name?

Reply:
I should leave an explanatory comment in the code. 

In some cases, it is necessary that in the absence of XmlDoc, x.Mark is called right on the spot so that specific nodes get into the declaration. For example, in the case of a primary constructor where an access modifier should be included in its declaration.